python/tests/mlir/quantum_type.py

```diff
@@ -0,0 +1,87 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+# RUN: PYTHONPATH=../../ pytest -rP  %s | FileCheck %s
+
+import pytest
+
+import cudaq
+
+def test_custom_quantum_type():
```

Comment:
@schweitzpgi It would be great to get your review on this again. I've updated to treat custom classes in Python that contain quantum memory as value types (`cc.struct` is allocated with `cc.undef` and values inserted and pulled with `cc.insertvalue / cc.extractvalue`. See the test FileCheck code below. 

Note the purely classical custom classes will continue using memory semantics (`cc.alloca` + `cc.load / cc.store`).